[not ready] Pycaffe improvements from OpenCL branch

Noiredd opened the pull request:
I'm bringing @naibaf7's pycaffe improvements from OpenCL branch to Master. Some pretty cool features have been implemented, most importantly exposure of `SolverParameter` and possibility to reinstatiate Solver, which would address #2766, and even update its parameters on-the-fly (#2686).  
I also want to expose `Solver::ApplyUpdate` (similar to how it's been done in [this fork](https://github.com/dosovits/caffe-fr-chairs/commit/6421df7fa8586364b843a52fd8ced8a1e8e2a813#diff-9b973c3db5bcd89fc0a5e1977f7cd70aR306)) to make it easier (possible at all?) to train GANs, but if it'd be better to have this in a separate PR (or done in a whole another way) - let me know.

Unfortunately many of the changes are interleaved with the evolution of that branch itself (modifications of `MemoryDataLayer`, adoption of `int_tp` type, some multi-device stuff, etc.) so it's not possible to just *pull them over to master* - plenty of conflicts arise due to interdependencies.  
This is a WIP with a request for feedback or ideas, especially from you @naibaf7 as the original author of those changes. I have tracked the history of OpenCL's pycaffe, dissecting every commit since July 2015, cherry-picking the relevant ones and cleaning them from OpenCL-specific code. Here is my plan of attack for this (ticked the already copied commits):
- [x] 61a05f3 `Python update, Lint fixes.` and d82979b `Softmax fix for >4 shape dims.`  
A heavy modification of the `MemoryDataLayer` and other stuff which I do not want to incorporate here; however, it exposes `Solver::max_iter` and allows passing a layer index in `Net_SetInputArrays`, so I just took the relevant pieces as a new commit.
- [x] faf003a `Pycaffe check update.` and 5ebebfb `Pycaffe fixed.`  
**ignored**, follow-ups to the `MemoryDataLayer` and `int_tp` changes
- [x] 01139b0 `Fix for Malis, Softmax.` and 165d231 `Removed duplicate function.`  
**ignored**, irrelevant to Master but introduce some conflicts around `.def("step")`
- [x] 984745d `Expose SolverParameter to Python.`  
This introduces an issue that I'm not sure how severe is nor how to deal with. Since it registers a new binding to `SolverParameter`, Python interpreter will now output the following warning every time we `import caffe`:  
`
/usr/local/caffe/python/caffe/pycaffe.py:13: RuntimeWarning: to-Python converter for caffe::SolverParameter already registered; second conversion method ignored.
  from ._caffe import Net, SGDSolver, NesterovSolver, AdaGradSolver, \
`
What do you think of this @naibaf7?
- [x] e2f8e70 `Improved pycaffe interface.`  
Changed meaning of `get_solver` will cause `make pytest` throw errors until 188aeae is merged.
- [x] e2f8e70 `Extended python interface.`  
**ignored**, requires some other commit for `Caffe::EnumerateDevices`
- [x] 52732be `Merge, stronger PyCaffe interface.`  
I thought this would also get rid of the `SolverParameter already registered` thing but it doesn't.
- [x] 97f9648 `Multi-device advancements.`
**ignored**, the change only relevant to OpenC's multi-device support
- [x] 03276d4 `updated pycaffe init` (adds `get_solver_from_file` to `__init__.py`)
- [x] 96b84b6 `Expose snapshot format to python` (I fixed a minor syntax error here)
- [ ] a35a577 `Python expose layers like blobs.` with its lint fixing commits, c65c85b and 6b88b6f
Adds `Net_SetLayerInputArrays` which is specific to `MemoryDataLayer` modifications so I'm not sure what to do with it. Probably it could be fitted with legacy accessors instead of `shape` to remove dependency on OpenCL's version of the layer. Thoughts?  
Also, this commit introduces a severe bug (changing `caffe.Net.layers` from `LayerVec` to `OrderedDict`).
- [ ] 8719549 `Pycaffe interface update.`  
Fixes the `Net.layers` bug but introduces lint errors which get fixed later in e61d542.
- [ ] a4c269e `Python interface extensions.`  
Exposes `NetParameter` and `NetState`, but also contains heavy changes to `Net_Init_Load` and `Net_SetLayerInputArrays`, so I won't merge this until I know what to do with a35a577.
- [ ] 188aeae `PyCaffe fixes for solver interface, ensure GIL lock in python layer.`  
Solves test problems introduced in e2f8e70, also changes behavior of the Python layer. Should we have that?
- [ ] 2daf478 `Python NetSpec extension for layer bundles` and its bugfix, 1d82ba2  
This should be ignored, right @naibaf7?
- [ ] e61d542 `LINT fixes.`  
A collection of fixes, mostly relevant due to errors in 8719549, but I'm planning to ignore this one and correct the lint manually within the related commit.
- [ ] 59113c0 `Removed duplicate layer parameter in PyCaffe.` (duplicated `LayerParameter`)
- [ ] 70858cb `PyCaffe fixes.`  
Lint errors which I would rather fix within a respective commit, but also changes `draw.py` - I'll look closer  whether they're needed.
- [ ] 77c3428 `Fix MSVC error for PyCaffe.` -- is this needed on master too?

All suggestions are welcome, but hints on dealing with 984745d, a35a577, 188aeae, 2daf478 and 77c3428 would be particularly helpful.

naibaf7 commented:
Looks good to me so far. I'll look over it in more detail after one of the Master branch maintainers had a peek.

Noiredd commented:
FYI: I pulled the remaining commits, sometimes changing them as I saw fit. Updated the description to reflect the changes, added a short summary below to ask about the issues I see right now.

shelhamer commented:
I will have a look this week! Thanks @Noiredd and @naibaf7

naibaf7 commented:
@shelhamer @Noiredd 
Actually I'm on my way to take this a lot further.
With my next update to OpenCL Caffe, there'll be:
- INT8, INT16, INT32, INT64 quantization
- HALF, FLOAT and DOUBLE precision
- Device abstractions
- Writing GPU layers for Caffe in Python [WIP for a future release]

And it will all be exposed to Python.
In the works here, if you want to see a preview of the new codebase:
https://github.com/naibaf7/caffe

naibaf7 commented:
Check the current progress here. It is now in a working state:
https://github.com/naibaf7/caffe/blob/master/python/caffe/_caffe.cpp

Noiredd commented:
@naibaf7 That is pretty spectacular! Thanks for giving a preview. Will the new numeric types be somehow documented? I did take a look but I'm not sure how to use them.

Also, does that render this PR pointless? I won't mind closing if your change will be incorporated in its place.

naibaf7 commented:
@Noiredd 
I think you should go ahead for the time being. My new branch will need 2-3 months longer to fully mature into "production ready".

Yes the numeric types will be documented. Float16 can already be used equally to float32 and float64, including unit tests (thanks to Intel, @gongzg, who started the float16 implementation; subsequently perfected by my efforts).
The integer (8, 16, 32, 64) quantization need some more testing and unit tests, I will check layer for layer on compatibility there.

Noiredd commented:
Update: I pulled all of the commits (assuming for now that 77c3428 is needed) and removed the old binding for `SolverParameter`. This PR is complete, albeit based on some assumptions.

Noiredd commented:
@shelhamer What's our status on this? Since we merged #6238, this causes some conflicts. Should I fix those and marge or is there an alternative plan?
